### ISSUE TITLE
[machine-runner] handle throws in reaction

### ISSUE THREAD

**Issue by Kelerchian** [Actyx/machines#75]
Currently, when a reaction throws an error, it is not handled gracefully.
When it happens, the runner simply prints the error and stops processing the events altogether.

The current implementation has an existing try-catch block that encapsulates the entire loop that iterates incoming events.
It suggests a natural evolution where errors are signaled to the machine's emitter, which then:

- The application code can listen to the error
- The error destroys the machine and throws any existing `for-await` loops, `next` call, and `peek` call in the user's code.

Let that be Evolution A.

However, with Zod support, unprocessable events (due to mismatched payload structure) are simply discarded, and regarded as invalid events. Reaction errors can be evolved into a similar approach instead, that is whenever a reaction throws, its events are regarded as invalid and are discarded. This approach will make reaction throw handling consistent with Zod parsing errors.

Let that be Evolution B 

Last but not least is to apply Evolution A but accompanied with a change in the Zod integration so that Zod parse failure results in the machine breaking, thus making it consistent with the new reaction handling.

Let that be Evolution C

#### Evolution A and Evolution C

```
let subscription = (events) => {
  try {
    for (let event of events) {
      internalStateMachine.consume(event)
    }
  } catch (error) {
    emitter.emit(error)
    rejectNext();
    destroyMachine();
  }
}
```

#### Evolution B 

```
let subscription = (events) => {
  for (let event of events) {
    try {
      internalStateMachine.consume(event)
    } catch (error) {
      emitter.emit(error)
    }
  }
}
```

### Implications

Notes:

- protocol here stands for the machine protocol and the swarm protocol
- history here means the chain of events that has been pre-recorded with any means (including manually writing to Actyx).

Implication on meaning:

- Evolution A and C rejects protocols that do not conform to written history.
- Evolution B rejects part of history that does not conform the protocol.

Eventuality of the different approach:

All evolutions verifies either the protocol or the event in history eventually. Rejecting a part of history eventually is acceptable, but rejecting a protocol eventually does not. A protocol is something agreed by multiple agents at design time, not runtime. If an agent can, in the middle of a runtime, decide that a protocol is not conforming, then it should not be there in the first place.

API design: 

Evolution A and C takes more effort to handle, yet is easier to control because it can be designed to involve the control flow (in the sense that, `for-await` breaks whenever the protocol is rejected).

Evolution B is harder to control because its only possible place as of the current API design is in the `error` hook. An alternative would be to have an optional config to destroy the machine (and possibly throw) whenever there is a non-matching history. Evolution B is the default implementation and evolution C becomes an optional feature switchable when calling the `createMachineRunner` function.

**Comment by rkuhn** [comment]
Another way to look at this is what meaning we should attribute to a reaction throwing:

1. we say that this signals a [failure](https://www.reactivemanifesto.org/glossary#Failure), or
2. we say that this signals an error.

In the first case, the machine is broken, which is your evolution A. In the second case, the event is considered erroneous, which fits your evolution B.

I’d rule out evolution C on the grounds that other nodes running different apps may emit events of unknown shape and tags, in which case resilience demands that we should simply ignore them — zod protects us from such crossfire.

---

Another angle is: given that we choose one of the two options, how can the programmer express the other choice? Our API forces the computation of a target state from a source state and a sequence of events that is determined by the state machine design. If the reaction logic determines that the given arguments constitute an error, then I don’t see a good way of fixing it, because we will only call the same function with the same arguments again later, arguably leading to the same error. I think this removes evolution B from our set of choices, leaving only evolution A.

**Comment by Kelerchian** [comment]
> If the reaction logic determines that the given arguments constitute an error, then I don’t see a good way of fixing it, because we will only call the same function with the same arguments again later, arguably leading to the same error.

Ah, there is a slight difference in the situation because I used the term 'throw'.
I need to emphasize that I refer to objects thrown both intentionally and unintentionally, especially thrown `TypeError` when accessing the property of undefined.
We can decide that throwing an error intentionally from a reaction is a bad practice (as we did with stashing commands).
Then the only errors remaining are `TypeError` and the likes (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError).

TypeError functions the same as Zod parse error, which drives this topic in the first place. I'd also leverage this option with your argument that different app may emit events of unknown shape and tags that cause this (and I assume this will only be fixed when the fleet management thing work in a super ideal condition).

Those facts combined would make more sense to approach it with Evolution B.

> Another angle is: given that we choose one of the two options, how can the programmer express the other choice?

With Evolution B as the default and no other features, Evolution A can be achieved:

```typescript
// user's application code

const machine = createSomeMachine()

let reactionError: null | MachineRunnerReactionException = null;
machine.on('error', (e) => {
  if (e instanceof MachineRunnerReactionException) {
    reactionError = e;
    machine.destroy()
  }
})

for await (const state of machine) (
}

```

**Comment by rkuhn** [comment]
We agree on demanding that throwing errors shall not be used by reaction functions explicitly.

---

Regarding the exceptions that remain, I think we agree that these are failures: the code is not equipped to deal with some given inputs. Now the question is whether this is caused by some differently shaped event emission of a remote node (in which case I agree we should ignore the event), or whether this is caused by a bug in the state handling function itself (in which case the machine is broken and cannot continue).

My vote is that zod support should be used to achieve the first part, and that remaining exceptions being thrown out of reaction handlers should terminate the machine. In the latter case, a `for await` loop shouldn’t just end but it should rethrow the exception to signal the breakage to calling code.

---

Summing up the above: using machine-check and zod guarantees that your reaction handlers will only be invoked with valid arguments, so any remaining exceptions are bugs. To me this sounds simple and intuitive.

**Comment by Kelerchian** [comment]
> that remaining exceptions being thrown out of reaction handlers should terminate the machine

With this, will be partial toward Zod, since we don't consider `TypeError` as the same (or at least, intersecting) class of error that should be handled the same way (although, it can be caused by bad programming as well)

This concern about being partial to Zod is that because the topic of runtime type parser/validator toolings within the community is now super dynamic in the sense that a library pops up every several months (for example https://www.reddit.com/r/typescript/comments/15v9jh6/typebox_a_high_performance_typescript_validator/).

---

In a sense, we are saying "Use Zod or face the consequences!".

IMO I will be very comfortable with evolution A if we can relax the runtime type parser/validator thing by externalizing it so that users can use any runtime type parser/validator. 

--- 

Another point to think about: we have the underutilized feature: `mock machine runner`, which is supposed to be used to contribute toward preventing bugs in reactions.

**Comment by rkuhn** [comment]
Good point about the mock runner: proper termination of the user state computations should always be covered by tests where types don’t cut it.

I find the distinction between faulty events (i.e. inputs) and faulty computation to be significant and intuitive. And you’re right in that we should support popular validator libraries. But we can already implement the intended semantics right now while Zod is the only choice. We’ll see how other validators may want to be integrated once we get corresponding user feedback.

**Comment by Kelerchian** [comment]
By the way, the custom verifier API I mentioned should look like this.
Zod and io-ts are simultaneously used in one API, although Zod will require a wrapper to work.

![image](https://github.com/Actyx/machines/assets/11693412/3abb1374-d0c4-4446-a9d1-198f8a6e13d1)

This opens a better way for collaborators to add support for any new and old validators because all they need to write is a wrapper like this.

**Comment by rkuhn** [comment]
nice!

**Comment by Kelerchian** [comment]
> nice!

Won't be done within this PR though. I'll make an issue for it if you're interested
